File: 0090-subsets-ii/0090-subsets-ii.java
class Solution {
    public List<List<Integer>> subsetsWithDup(int[] nums) {
        List<List<Integer>> anw = new ArrayList<>();
        List<Integer> list = new ArrayList<>();
        Arrays.sort(nums);
        helper(0, nums, anw, list);
        return anw;
    }

    private void helper(int start, int[] nums, List<List<Integer>> anw, List<Integer> list) {
        if (start == nums.length) {
            List<Integer> temp = new ArrayList<>(list);
            anw.add(temp);
            return;
        }
        list.add(nums[start]);
        helper(start + 1, nums, anw, list);
        list.remove(list.size() - 1);

        while (start < nums.length - 1 && nums[start] == nums[start + 1]) start++;
        helper(start + 1, nums, anw, list);
    }
}

File: 0090-subsets-ii/0090-subsets-ii.cpp
class Solution {
public:
    vector<vector<int>> subsetsWithDup(vector<int>& nums) {
        std::sort(nums.begin(), nums.end());
        vector<vector<int>> res;
        vector<int> curr;
        dfs(nums, 0, res, curr);
        return res;
    }

    void dfs(vector<int>& nums, size_t i, vector<vector<int>>& res, vector<int>& curr) {
        if (i >= nums.size()) {
            res.push_back(curr);
            return;
        }

        curr.push_back(nums[i]);
        dfs(nums, i + 1, res, curr);
        curr.pop_back();
        while(i >= 0 && i < nums.size() - 1 && nums[i] == nums[i + 1]) i++;
        dfs(nums, i + 1, res, curr);
    }
};